drivers/usbdev_synopsys_dwc2: add GD32V support

gschorcht opened the pull request:
### Contribution description

This PR provides the GD32V support in `drivers/usbdev_synopsys_dwc2`. It also includes some cleanup and bug fixes.

To support GD32V (CID 1.000) the PR includes the following changes:
- XFRC interrupts are also used in Non-DMA mode to complete a transfer
- SETUP phase done (STUP) interrupt is used in Non-DMA mode
- additional XFRC interrupts in SETUP stage are ignored
- RX FIFO handling doesn't complete a transfer anymore and is handled directly in ISR
- OTG interrupt handling added

The following fixes, improvements and cleanup has been added:
- USB OTG HS definition dependencies for platforms that don't support USB OTG HS
- thread context switch at ISR exit was added for all platforms
- units in USB FIFO size definitions are clarified
- USB EP number as defined in CMSIS is used
- FIFO allocation documentation is improved
- power modes for STM32 are fixed
- CID/HW debug info added

The PR could be split into a PRs with cleanups and improvements and a PR with GD32V support if necessary.

### Testing procedure
```
BOARD=sipeed-longan-nano make -j8 -C tests/shell
```
should work with `stdio_cdc_acm`.

`periph_usbdev` should still work for other platforms. The PR was tested with `tests/usbus_cdc_ecm` together with `stdio_cdc_acm` if possible:
```
esp32s2      FS  CID: 0000, HWREV: 4f54400a, HWCFG: 224dd930  (esp32s2-devkit)
esp32s3      FS  CID: 0000, HWREV: 4f54400a, HWCFG: 224dd930  (esp32s3-devkit)
efm32gg12    FS  CID: 0000, HWREV: 4f54330a, HWCFG: 228f5910  (sltb009a)
gd32vf103cb  FS  CID: 1000, HWREV: 00000000, HWCFG: 00000000  (sipeed-longan-nano)
stm32f407vg  FS  CID: 1200, HWREV: 4f54281a, HWCFG: 229dcd20  (stm32f4discovery) 
stm32f429zi  HS  CID: 1100, HWREV: 4f54281a, HWCFG: 229ed590  (stm32f429i-disc1)
stm32f439zi  FS  CID: 1200, HWREV: 4f54281a, HWCFG: 229dcd20  (nucleo-f439ze)
stm32f446ze  FS  CID: 2000, HWREV: 4f54320a, HWCFG: 229ed520  (nucleo-f446ze)
stm32f723ie  FS  CID: 3000, HWREV: 4f54330a, HWCFG: 229ed520  (stm32f723e-disco)
stm32f723ie  HS  CID: 3100, HWREV: 4f54330a, HWCFG: 229fe1d0  (stm32f723e-disco)
stm32f746ng  FS  CID: 2000, HWREV: 4f54320a, HWCFG: 229ed520  (stm32f746g-disco)
stm32f746ng  HS  CID: 2100, HWREV: 4f54320a, HWCFG: 229fe190  (stm32f746g-disco)
stm32f767zi  FS  CID: 2000, HWREV: 4f54320a, HWCFG: 229ed520  (nucleo-f767zi)
```
Boards with HS interfaces were tested in DMA and none-DMA mode.

### Issues/PRs references

benpicco reviewed (approved):
Works like a charm and I see you already have tested this on a wide range of implementations.

Please squash.

benpicco commented on drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:
```diff
@@ -1444,27 +1590,29 @@ void _isr_common(dwc2_usb_otg_fshs_t *usbdev)
 {
     const dwc2_usb_otg_fshs_config_t *conf = usbdev->config;
 
-    uint32_t status = _global_regs(conf)->GINTSTS;
+    uint32_t status = _global_regs(conf)->GINTSTS & _global_regs(conf)->GINTMSK;
+
+    _global_regs(conf)->GAHBCFG &= ~USB_OTG_GAHBCFG_GINT;
 
     if (status) {
-        if ((status & USB_OTG_GINTSTS_RXFLVL) && !_uses_dma(conf)) {
-            unsigned epnum = _global_regs(conf)->GRXSTSR &
-                             USB_OTG_GRXSTSP_EPNUM_Msk;
-            usbdev->usbdev.epcb(&usbdev->out[epnum].ep, USBDEV_EVENT_ESR);
+        if (status & USB_OTG_GINTSTS_RXFLVL) {
+            /* Only in non-DMA mode the RXFLVL interrupt is enabled and used
+             * to read out the RX FIFO directly in the ISR. */
+            _read_packet(usbdev);
+            _global_regs(conf)->GAHBCFG |= USB_OTG_GAHBCFG_GINT;
         }
-        else if (_global_regs(conf)->GINTSTS &
-                 (USB_OTG_GINTSTS_OEPINT | USB_OTG_GINTSTS_IEPINT)) {
+        else if (status & (USB_OTG_GINTSTS_OEPINT | USB_OTG_GINTSTS_IEPINT)) {
             _isr_ep(usbdev);
         }
         else {
             /* Global interrupt */
             usbdev->usbdev.cb(&usbdev->usbdev, USBDEV_EVENT_ESR);
         }
-        _global_regs(conf)->GAHBCFG &= ~USB_OTG_GAHBCFG_GINT;
     }
-#ifdef MCU_STM32
-    cortexm_isr_end();
-#endif
+
+    if (sched_context_switch_request) {
+        thread_yield_higher();
+    }
```
Turns out this is already handled by [common code](https://github.com/RIOT-OS/RIOT/blob/master/cpu/riscv_common/irq_arch.c#L195-L202) on RISC-V so this is only needed on Cortex-M

gschorcht replied:
OK, I see. For ESP32x it should also part of ISR exit code.

gschorcht commented:
> Works like a charm and I see you already have tested this on a wide range of implementations.

I would like to have look from @bergzand onto the driver since its interrupt handling has been changed completely (commits bcc8c2beb4455816860b230bb4d7dbdffaa9a38b, 38620fd88a745e13630fd8d60d0a0140b8f9bed0, ad191b14ea9ce85ebd6c0397d6268a87042419e9)

bergzand commented:
> I would like to have look from @bergzand onto the driver since its interrupt handling has been changed completely (commits [bcc8c2b](https://github.com/RIOT-OS/RIOT/commit/bcc8c2beb4455816860b230bb4d7dbdffaa9a38b), [38620fd](https://github.com/RIOT-OS/RIOT/commit/38620fd88a745e13630fd8d60d0a0140b8f9bed0), [ad191b1](https://github.com/RIOT-OS/RIOT/commit/ad191b14ea9ce85ebd6c0397d6268a87042419e9))

I see one potential race condition here, but not sure if it is the case. For the non-DMA case both the rxfifolvl and the transfer complete interrupts are enabled. I previously had a similar issue with the DMA case where both the transfer complete and the rxfifolvl interrupts were enabled (see #18726). On fast microcontrollers with HS phy this could trigger unstable behaviour quickly when using cdc ecm and pinging the device with large icmp packets. I wonder if this issue returns here but for the non-dma case, with a HS phy.

I'm also curious why a different interrupt is used during the address state.

gschorcht commented:
> I see one potential race condition here, but not sure if it is the case. For the non-DMA case both the rxfifolvl and the transfer complete interrupts are enabled.

The causal order should be that `RXFLVL` comes before `XFRC`. There shouldn't be a race condition because only one interrupt is handled per ISR call and `RXFLVL` is always handled before `XFRC` even if `RXFLVL` and `XFRC` are triggered at the same time. In addition, `RXFLVL` is handled synchronously directly in the ISR while `XFRC` is handled asynchronously using thread events. Also, `RXFLVL` is only used to copy the FIFO contents and to set a flag indicating that a `SETUP` transfer has been started. The latter is seems to be needed certain core versions such as the GD32V core version 1.000.

> I'm also curious why a different interrupt is used during the address state.

Because TXFE interrupt doesn't seem to work in default state in core version 1.000 :worried:

gschorcht commented on drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:
```diff
@@ -587,21 +648,32 @@ static void _flush_rx_fifo(const dwc2_usb_otg_fshs_config_t *conf)
-     * while (un)blocking works on the stm32f401, needs more
-     * investigation with a larger set of chips */
-#if defined(STM32_USB_OTG_CID_1x)
+#if defined(MCU_STM32)
```
@bergzand I'm not sure whether this change could cause not to work on stm32f446 any longer. I have tested it with a number of non CID 1.x core versions, but couldn't find any problem.

gschorcht commented:
I had to rebase to resolve the conflicts.

bergzand commented:
Tested this both on the gd32v (sipeed longan nano) and on an stm32f446re with the usb-otg-hs peripheral using an external phy to achieve HS USB speed.

The only issue I could find is that my board doesn't enumerate anymore when I disable DMA by always return `false` in the `_uses_dma()` function.

bergzand commented on drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c:
```diff
@@ -749,10 +807,53 @@ static void _usbdev_init(usbdev_t *dev)
+    /* ensure that PLL clock is used */
+    assert((RCU->CFG0 & RCU_CFG0_SCS_Msk) == RCU_CFG0_SCS_PLL);
+    /* ensure that the 48 MHz USB clock can be generated */
+    assert((CLOCK_CORECLOCK == MHZ(48)) || (CLOCK_CORECLOCK == MHZ(72)) ||
+           (CLOCK_CORECLOCK == MHZ(96)) || (CLOCK_CORECLOCK == MHZ(120)));
```
Could this be changed into a `static_assert`?

gschorcht replied:
Yep, changed.

bergzand reviewed (changes requested):
Not much, most changes look sane. Thanks again for picking this up.

see #19471 for the comments on the set address bits

bergzand commented:
Please rebase now that #19471 is merged

gschorcht commented:
> Please rebase now that #19471 is merged

Thanks for merging PR #19471. I rebased the PR and squashed the small change with static assert right away.

benpicco commented:
bors try

bergzand reviewed (approved):
Ack, thanks for the many improvements!

bors merge

gschorcht commented:
Thanks for reviewing, testing und merging.
